src/main/drivers/sound_beeper.c

```diff
@@ -24,43 +24,55 @@
 #include "io.h"
 
 #include "sound_beeper.h"
+#include "pwm_output.h"
 
 
 #ifdef BEEPER
-
 static IO_t beeperIO = DEFIO_IO(NONE);
 static bool beeperInverted = false;
-
+static uint16_t beeperFrequency = 0;
 #endif
 
+
 void systemBeep(bool onoff)
 {
-#ifndef BEEPER
-    UNUSED(onoff);
+#ifdef BEEPER
+    if(beeperFrequency == 0) {
+        IOWrite(beeperIO, beeperInverted ? onoff : !onoff);
+    } else {
+        pwmWriteBeeper(onoff);
+    }
 #else
-    IOWrite(beeperIO, beeperInverted ? onoff : !onoff);
+    UNUSED(onoff);
 #endif
 }
 
 void systemBeepToggle(void)
 {
 #ifdef BEEPER
-    IOToggle(beeperIO);
+    if(beeperFrequency == 0) {
+        IOToggle(beeperIO);
+    } else {
+        pwmToggleBeeper();
+    }
 #endif
 }
 
 void beeperInit(const beeperDevConfig_t *config)
 {
-#ifndef BEEPER
-    UNUSED(config);
-#else
-    beeperIO = IOGetByTag(config->ioTag);
-    beeperInverted = config->isInverted;
-
-    if (beeperIO) {
-        IOInit(beeperIO, OWNER_BEEPER, 0);
-        IOConfigGPIO(beeperIO, config->isOpenDrain ? IOCFG_OUT_OD : IOCFG_OUT_PP);
+#ifdef BEEPER
+    beeperFrequency = config->frequency;
+    if(beeperFrequency == 0) {
+        beeperIO = IOGetByTag(config->ioTag);
+        beeperInverted = config->isInverted;
+        if (beeperIO) {
+            IOInit(beeperIO, OWNER_BEEPER, 0);
+            IOConfigGPIO(beeperIO, config->isOpenDrain ? IOCFG_OUT_OD : IOCFG_OUT_PP);
+        }
+    } else {
+        beeperPwmInit(beeperFrequency);
     }
-    systemBeep(false);
```

Comment:
Sorry, I wasn't clear enough. `systemBeep(false)` should not be completely removed. It should be called outside of the `else` block, like it was in the original code.